### PR TITLE
integration: fix imagestream admission flake

### LINKED ISSUE #9775: Unable to deploy database in e2e test: secret never got mounted
Post Rebase (kube 1.3.0) on our RHEL 7.2 CI machines some deployments are failing because they are unable to mount a secret occasionally.

https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/3091/
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/3091/s3/download/test-end-to-end-docker/logs/container-origin.log

Opened kubernetes/kubernetes#28750

### LINKED ISSUE #9780: Integration flake: TestImageStreamAdmitStatusUpdate
https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/3118/

```
[294.303731ms] [132.502µs] END
I0711 13:42:20.191552    2472 ensure.go:86] Added service-load-balancer-controller service accounts to the system:service-load-balancer-controller cluster role: <nil>
I0711 13:42:20.264538    2472 clusterquotamapping.go:302] Waiting for the caches to sync before starting the quota mapping controller workers
I0711 13:42:20.350323    2472 ensure.go:86] Added service-serving-cert-controller service accounts to the system:service-serving-cert-controller cluster role: <nil>
I0711 13:42:20.364886    2472 clusterquotamapping.go:302] Waiting for the caches to sync before starting the quota mapping controller workers
W0711 13:42:20.402063    2472 run_components.go:201] Binding DNS on port 8053 instead of 53, which may not be resolvable from all clients
I0711 13:42:20.402385    2472 reflector.go:213] Starting reflector *api.Service (30m0s) from github.com/openshift/origin/pkg/dns/serviceaccessor.go:45
I0711 17:42:20.402569Z 2472 log.go:289] 2016-07-11T17:42:20Z skydns: ready for queries on cluster.local. for tcp4://127.0.0.1:8053 [rcache 0]
I0711 17:42:20.402596Z 2472 log.go:289] 2016-07-11T17:42:20Z skydns: ready for queries on cluster.local. for udp4://127.0.0.1:8053 [rcache 0]
I0711 13:42:20.402739    2472 reflector.go:253] Listing and watching *api.Service from github.com/openshift/origin/pkg/dns/serviceaccessor.go:45
I0711 13:42:20.466511    2472 net.go:105] Got error &net.OpError{Op:"dial", Net:"tcp", Source:net.Addr(nil), Addr:(*net.TCPAddr)(0xc828deda70), Err:(*os.SyscallError)(0xc8262d3500)}, trying again: "127.0.0.1:8053"
I0711 13:42:20.487899    2472 clusterquotamapping.go:302] Waiting for the caches to sync before starting the quota mapping controller workers
I0711 13:42:20.566971    2472 run_components.go:218] DNS listening at 127.0.0.1:8053
I0711 13:42:20.567028    2472 reflector.go:202] Starting reflector *api.Namespace (2m0s) from github.com/openshift/origin/pkg/project/auth/cache.go:186
I0711 13:42:20.570851    2472 reflector.go:213] Starting reflector *api.ClusterResourceQuota (10m0s) from github.com/openshift/origin/pkg/controller/shared/shared_informer.go:79
I0711 13:42:20.570892    2472 reflector.go:253] Listing and watching *api.ClusterResourceQuota from github.com/openshift/origin/pkg/controller/shared/shared_informer.go:79
I0711 13:42:20.571202    2472 reflector.go:253] Listing and watching *api.Namespace from github.com/openshift/origin/pkg/project/auth/cache.go:186
I0711 13:42:20.572982    2472 reflector.go:213] Starting reflector *api.Namespace (10m0s) from github.com/openshift/origin/pkg/controller/shared/shared_informer.go:79
I0711 13:42:20.573015    2472 reflector.go:253] Listing and watching *api.Namespace from github.com/openshift/origin/pkg/controller/shared/shared_informer.go:79
I0711 13:42:20.613087    2472 clusterquotamapping.go:107] Starting workers for quota mapping controller workers
--- FAIL: TestImageStreamAdmitStatusUpdate (6.22s)
	imagestream_admission_test.go:588: creating limit range object "limits" with openshift.io/ImageStream limited to: map[openshift.io/image-tags:{{0 0} {<nil>} 0 DecimalSI} openshift.io/images:{{0 0} {<nil>} 0 DecimalSI}]
	imagestream_admission_test.go:383: trying to create a new image stream with a tag exceeding limit map[openshift.io/images:{{0 0} {<nil>} 0 DecimalSI} openshift.io/image-tags:{{0 0} {<nil>} 0 DecimalSI}]
	imagestream_admission_test.go:393: adding new tag to image stream status exceeding limit map[openshift.io/image-tags:{{0 0} {<nil>} 0 DecimalSI} openshift.io/images:{{0 0} {<nil>} 0 DecimalSI}]
	imagestream_admission_test.go:597: bump a limit on resource "openshift.io/images" to 1
	imagestream_admission_test.go:422: adding new tag to image stream status below limit map[openshift.io/image-tags:{{0 0} {<nil>} 0 DecimalSI} openshift.io/images:{{0 0} {<nil>} 0 DecimalSI}]
	imagestream_admission_test.go:440: unexpected error: ImageStream "is" is invalid: imageStream: Forbidden: ImageStream "is" is forbidden: requested usage of openshift.io/images exceeds the maximum limit per openshift.io/ImageStream (1 > 0)
FAIL
Running TestImageStreamCreate...
```

### PR DESCRIPTION
**Pull request by miminar**
Retry on quota exceeded after incrementing limits. The admission plugin
may hit a cache with outdated value.

Closes #9780

### CODE REVIEW

**Comment by deads2k** [comment]
If the problem is in an admission plugin that isn't handling a conflict failure on update, you should probably update that admission plugin to retry like we do for quota.

**Comment by deads2k** [comment]
@smarterclayton want to generalize the Update cache layer?  Its easy for direct gets, its a lot more expensive for `List` operations since you have to walk the list.

**Comment by miminar** [comment]
> If the problem is in an admission plugin that isn't handling a conflict failure on update, you should probably update that admission plugin to retry like we do for quota.

@deads2k actually, the conflict can happen only during an update of image stream, not the limits. The problem is a cache (indexer) used in [limitVerifier](https://github.com/openshift/origin/blob/master/pkg/image/admission/imagestream_limits.go#L43) that may be outdated right after the limits are bumped. From my understanding it's not that harmful to act on 1 second old quota or limit range. But I may be wrong.

**Comment by deads2k** [comment]
> From my understanding it's not that harmful to act on 1 second old quota or limit range. But I may be wrong.

That is true.  Do you know what is conflicting within this test?  If its an unrelated istag, you should be able to detect that in the RESTStorage and retry if it has no effect on this particular update.

**Comment by miminar** [comment]
> Do you know what is conflicting within this test?

@deads2k Not sure, what has changed but I couldn't reproduce a single conflict in 400 runs locally. I've removed retries on conflict. Tests now retry only on outdated limits in cache.

**Comment by miminar** [comment]
@deads2k would you mind another look?

File: test/integration/imagestream_admission_test.go [hunk 1]
```diff
@@ -634,7 +616,7 @@ func retryOnQuotaExceeded(t *testing.T, retries int, fn func() error) error {
 			return err
 		}
 		t.Logf("got quota exceeded error, retrying after sleep; number of retries remaining: %d", retries-i-1)
-		time.Sleep(time.Second)
+		time.Sleep(time.Millisecond * 500)
```
Comment (deads2k) [inline test/integration/imagestream_admission_test.go]: why?

File: test/integration/imagestream_admission_test.go [hunk 2]
```diff
@@ -86,7 +86,7 @@ func TestImageStreamTagsAdmission(t *testing.T) {
 		},
 	}
 	// we may hit cache with old limit, let's retry in such a case
-	err = retryOnQuotaExceeded(t, 1, func() error {
+	err = retryOnQuotaExceeded(t, 3, func() error {
```
Comment (deads2k) [inline test/integration/imagestream_admission_test.go]: Implementation of this method should be updated to use `wait.ExponentialBackoff`.  If you think its cache latency, then we should be able to hit it again quickly.

Also, the old impl of passing `1` would never have retried, right?

Comment (miminar) [inline test/integration/imagestream_admission_test.go]: > Implementation of this method should be updated to use wait.ExponentialBackoff. If you think its cache latency, then we should be able to hit it again quickly.

Yes, I'll rewrite.

> Also, the old impl of passing 1 would never have retried, right?

That's right. It made me laugh until I realized it's my code...

File: test/integration/imagestream_admission_test.go (hunk 1 above)
Comment (miminar) [inline test/integration/imagestream_admission_test.go]: I find 1 second too long to wait. Moreover, there was no retry before. Now we have two, which I'd like to bump to three.

**Comment by miminar** [comment]
Comments addressed. Thanks!

**Comment by deads2k** [comment]
[merge]

**Pull request merged**